sse2neon.h

```diff
@@ -8358,6 +8359,649 @@ FORCE_INLINE int _mm_testz_si128(__m128i a, __m128i b)
+                                            int lb)
+{
+    __m128i mtx[16];
+    PCMPSTR_EQ_8x16(a, b, mtx);
+    return aggregate_equal_ordered_8x16(16, la, lb, mtx);
+}
+
+static int _sse2neon_cmp_word_equal_ordered(__m128i a,
+                                            int la,
+                                            __m128i b,
+                                            int lb)
+{
+    __m128i mtx[16];
+    PCMPSTR_EQ_16x8(a, b, mtx);
+    return aggregate_equal_ordered_16x8(8, la, lb, mtx);
+}
+
+typedef int (*cmpestr_func_t)(__m128i a, int la, __m128i b, int lb);
+
+static cmpestr_func_t _sse2neon_cmpfunc_table[] = {
+    /* CMP_UBYTE_EQUAL_ANY */ _sse2neon_cmp_byte_equal_any,
+    /* CMP_UWORD_EQUAL_ANY */ _sse2neon_cmp_word_equal_any,
+    /* CMP_SBYTE_EQUAL_ANY */ _sse2neon_cmp_byte_equal_any,
+    /* CMP_SWORD_EQUAL_ANY */ _sse2neon_cmp_word_equal_any,
+    /* CMP_UBYTE_RANGES */ _sse2neon_cmp_ubyte_ranges,
+    /* CMP_UWORD_RANGES */ _sse2neon_cmp_uword_ranges,
+    /* CMP_SBYTE_RANGES */ _sse2neon_cmp_sbyte_ranges,
+    /* CMP_SWORD_RANGES */ _sse2neon_cmp_sword_ranges,
+    /* CMP_UBYTE_EQUAL_EACH */ _sse2neon_cmp_byte_equal_each,
+    /* CMP_UWORD_EQUAL_EACH */ _sse2neon_cmp_word_equal_each,
+    /* CMP_SBYTE_EQUAL_EACH */ _sse2neon_cmp_byte_equal_each,
+    /* CMP_SWORD_EQUAL_EACH] */ _sse2neon_cmp_word_equal_each,
+    /* CMP_UBYTE_EQUAL_ORDERED */ _sse2neon_cmp_byte_equal_ordered,
+    /* CMP_UWORD_EQUAL_ORDERED */ _sse2neon_cmp_word_equal_ordered,
+    /* CMP_SBYTE_EQUAL_ORDERED */ _sse2neon_cmp_byte_equal_ordered,
+    /* CMP_SWORD_EQUAL_ORDERED */ _sse2neon_cmp_word_equal_ordered};
+
+static inline int neg_fun(int res, int lb, int imm8, int bound)
+{
+    switch (imm8 & 0x30) {
+    case _SIDD_NEGATIVE_POLARITY:
+        res ^= 0xffffffff;
+        break;
+    case _SIDD_MASKED_NEGATIVE_POLARITY:
+        res ^= (1 << lb) - 1;
+        break;
+    default:
+        break;
+    }
+
+    return res & ((bound == 8) ? 0xFF : 0xFFFF);
+}
+
+// Compare packed strings in a and b with lengths la and lb using the control
+// in imm8, and store the generated index in dst.
+// https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=_mm_cmpestri
+FORCE_INLINE int _mm_cmpestri(__m128i a,
+                              int la,
+                              __m128i b,
+                              int lb,
+                              const int imm8)
+{
+    int upper = (imm8 & 0x01) ? 8 : 16;
+#if defined(__aarch64__)
+    __asm__ __volatile__(
+        "eor w0, %w[a], %w[a], asr #31\n"
+        "sub %w[a], w0, %w[a], asr #31\n"
+        "eor w1, %w[b], %w[b], asr #31\n"
+        "sub %w[b], w1, %w[b], asr #31\n"
+        "cmp %w[a], %w[bd]\n"
+        "csel %w[a], %w[bd], %w[a], gt\n"
+        "cmp %w[b], %w[bd]\n"
+        "csel %w[b], %w[bd], %w[b], gt\n"
+        : [a] "+r"(la), [b] "+r"(lb)
+        : [bd] "r"(upper)
+        : "w0", "w1");
+#else
+    __asm__ __volatile__(
```

Comment:
The instruction `csel` is essentially like the ternary operator in C, and it is absent in ARMv7-A. Thus, we can have C implementation for Arm32 rather than assembly listing.

Reply:
For my curiosity, what's the reason not to use inline assembly but use C implementation instead in ARMv7-A? Currently I can only come up with the idea of best portability between ARM and thumb instruction set.

Reply:
If we implement in generic C, we can benefit from compiler optimizations without paying attention to arm-thumb interworking.